tree/node.go

```diff
@@ -0,0 +1,101 @@
+package tree
+
+import (
+	"github.com/sasswart/gin-in-a-can/errors"
+)
+
+var _ NodeTraverser = &Node{}
+
+type Metadata map[string]string
+type Node struct {
+	basePath string
+	parent   NodeTraverser
+	metadata Metadata
+	Name     string
+}
+type Noder interface {
+	SetName(name string)
+	GetName() []string
+
+	GetBasePath() string
+	SetBasePath(path string)
+
+	GetRef() string
+	SetMetadata(metadata Metadata)
+	GetMetadata() Metadata
+}
+
+//
+//// Noder Functions
+//
+
+func (n *Node) SetName(name string) {
+	n.Name = name
+}
+func (n *Node) GetName() []string {
+	if n.GetParent() == nil {
+		return []string{n.Name}
+	}
+	return append(n.GetParent().GetName(), n.Name)
+}
+
+func (n *Node) SetBasePath(path string) {
```

Comment:
Am I correct that any ref-having Noder will override SetBasePath? If so, can we document that? Such distributed contracts between types aren't obvious to newcomers

Reply:
Ah. BasePath is always from the root of the tree, but the ref is concatenated on when resolving refs. Took me a sec.